calico_containers/pycalico/block.py

```diff
@@ -0,0 +1,383 @@
+BLOCK_SIZE = 2 ** BLOCK_SIZE_BITS
+PREFIX_MASK = {4: (IPAddress("255.255.255.255") ^ (BLOCK_SIZE - 1)),
+               6: (IPAddress("ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff") ^
+                   (BLOCK_SIZE - 1))}
+my_hostname = socket.gethostname()
+
+
+
+class AllocationBlock(object):
+    """
+    A block of IP addresses from which to allocate for IPAM clients.
+
+    Blocks are identified by IP prefix.  Each block is a single, keyed object
+    in etcd and the value of the block object in the datastore encodes all the
+    allocations for all the IP addresses within that prefix.
+
+    Thus, allocations and releases of IP addresses correspond to changes in the
+    block's value.  Compare-and-swap atomicity is used to ensure allocations
+    and releases are consistent operations.
+    """
+    CIDR = "cidr"
+    AFFINITY = "affinity"
+    HOST_AFFINITY_T = "host:%s"
+    ALLOCATIONS = "allocations"
+    ATTRIBUTES = "attributes"
+    ATTR_PRIMARY = "primary"
+    ATTR_SECONDARY = "secondary"
+
+    def __init__(self, cidr_prefix, host_affinity):
+        assert isinstance(cidr_prefix, IPNetwork)
+        assert cidr_prefix.cidr == cidr_prefix
+
+        # Make sure the block is the right size.
+        assert cidr_prefix.prefixlen == (BITS_BY_VERSION[cidr_prefix.version] -
+                                         BLOCK_SIZE_BITS)
+        self.cidr = cidr_prefix
+        self.db_result = None
+
+        self.host_affinity = host_affinity
+        """
+        Both to minimize collisions, where multiple hosts attempt to change a
+        single block, and to support route aggregation, each block has affinity
+        to a single Calico host.  That host does not hold exclusive rights to
+        modify the block; any host may still do that.  The host with affinity
+        simply uses the block as the place where it first searches if the user
+        asked to have the IP assigned automatically.
+        """
+
+        self.allocations = [None] * BLOCK_SIZE
+        """
+        A fixed length array with one entry for every address in the block.
+        None means unallocated.  A non-negative integer indicates the address
+        is allocated, and is the index into the `attributes` array for the
+        attributes assigned to the allocation.
+        """
+
+        self.attributes = []
+        """
+        Dictionaries of attributes for allocations.
+        """
+
+    def to_json(self):
+        """
+        Convert to a JSON representation for writing to etcd.
+        """
+
+        json_dict = {AllocationBlock.CIDR: str(self.cidr),
+                     AllocationBlock.AFFINITY:
+                         AllocationBlock.HOST_AFFINITY_T % self.host_affinity,
+                     AllocationBlock.ALLOCATIONS: self.allocations,
+                     AllocationBlock.ATTRIBUTES: self.attributes}
+        return json.dumps(json_dict)
+
+    @classmethod
+    def from_etcd_result(cls, etcd_result):
```

Comment:
from_json would be nicer I'd have thought... it mirrors the the to_json method, and doesn't tie this to an etcd backend.